// @shopify/ui-extensions-react/src/surfaces/customer-account.ts
export {
  extend,
  extension,
  isConditionalStyle,
  isConditionalStyleWithDefault,
  Style,
} from '@shopify/ui-extensions/customer-account';
export {render, reactExtension} from './customer-account/render';

export type {
  AccessibilityRole,
  ActionExtensionApi,
  Alignment,
  AllowedComponentsForRenderExtension,
  ApiForRenderExtension,
  Appearance,
  AtLeastOne,
  ArgumentsForExtension,
  Autocomplete,
  AutocompleteField,
  AutocompleteGroup,
  Background,
  BackgroundFit,
  BackgroundPosition,
  BackgroundProps,
  BackgroundRepeat,
  BlockAlignment,
  BorderRadius,
  BorderProps,
  CornerProps,
  CornerRadius,
  BorderStyle,
  BorderWidth,
  Breakpoint,
  Columns,
  Conditions,
  ConditionalStyle,
  ConditionalValue,
  DayString,
  DateString,
  DateRange,
  DisclosureActivatorProps,
  DisclosureOpen,
  DisabledDate,
  ExtensionTarget,
  ExtensionTargets,
  Fit,
  FullExtensionNavigation,
  FullPageApi,
  GridItemSize,
  I18n,
  I18nTranslate,
  InlineAlignment,
  MaybeConditionalStyle,
  MaybeResponsiveConditionalStyle,
  InteractiveConditions,
  Language,
  Localization,
  MaybeShorthandProperty,
  NonPresentationalAccessibilityRole,
  OverlayActivatorProps,
  PopoverPosition,
  RenderExtension,
  RenderExtensions,
  ResolutionCondition,
  ReturnTypeForExtension,
  Rows,
  ScrollViewEvent,
  ShopifyGlobal,
  ShorthandProperty,
  SelectedDate,
  Size,
  SizingProps,
  Spacing,
  SpacingProps,
  Status,
  Version,
  VisibilityProps,
  ViewCoordinate,
  ViewTranslate,
  ViewportSizeCondition,
  ViewPosition,
  ViewPositionType,
  ViewLikeAccessibilityRole,
  YearMonth,
} from '@shopify/ui-extensions/customer-account';

export * from './customer-account/components';
export * from './customer-account/hooks';
